docs/architecture/adr-017-historical-header-module.md

```diff
@@ -0,0 +1,59 @@
+# ADR 17: Historical Header Module
+
+## Changelog
+
+- 26 November 2019: Start of first version
+- 2 December 2019: Final draft of first version
+
+## Context
+
+In order for the Cosmos SDK to implement the [IBC specification](https://github.com/cosmos/ics), modules within the SDK must have the ability to introspect recent consensus states (validator sets & commitment roots) as proofs of these values on other chains must be checked during the handshakes.
+
+## Decision
+
+The application MUST store the most recent `n` headers in a persistent store. At first, this store MAY be the current Merklised store. A non-Merklised store MAY be used later as no proofs are necessary.
+
+The application MUST store this information by storing new headers immediately when handling `abci.RequestBeginBlock`:
+
+```golang
+func BeginBlock(ctx sdk.Context, keeper HistoricalHeaderKeeper, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
```

Comment:
is this part of a new module?

Reply:
It can be. If we do anything similar in another module perhaps we could combine them.

Reply:
Added a comment about this.